File: codeforces/702E/main.cc
// ID:  702E (Analysis of Pathes in Functional Graph)
// URL: http://codeforces.com/problemset/problem/702/E

#include <stdio.h>
#include <iostream>
#include <algorithm>
using namespace std;

const int MAX = 100000;
int weight[MAX], conn[MAX];

// dp[i][j].to:     starting from node i, after 2^j iterations, where am I?
// dp[i][j].cost:   path's cost
// dp[i][j].weight: minimum weight in the path
struct node {
	int to;
	long long int cost;
	int minWeight;
};
node dp[MAX][60];

int main() {
	int n;
	long long int goal;

	// input
	cin >> n >> goal;
	for(int i=0; i<n; i++) scanf("%d", conn + i);
	for(int i=0; i<n; i++) scanf("%d", weight + i);

	// DP: base case
	for(int i=0; i<n; i++){
		dp[i][0].to = conn[i];
		dp[i][0].cost = weight[i];
		dp[i][0].minWeight = weight[i];
	}

	// DP: recursive case
	for(int j=1; j<60; j++){
		for(int i=0; i<n; i++){
			int mid = dp[i][j-1].to;
			dp[i][j].to = dp[mid][j-1].to;
			dp[i][j].cost = dp[i][j-1].cost + dp[mid][j-1].cost;
			dp[i][j].minWeight = min(dp[i][j-1].minWeight, dp[mid][j-1].minWeight);
		}
	}

	// output
	for(int i=0; i<n; i++){
		// print s_i and m_i
		long long int k = goal;
		int j = 0;
		long long int s = 0;
		int m = 1e8 + 1;
		int from = i;
		while(k){
			if(k & 1){
				s += dp[from][j].cost;
				m = min(m, dp[from][j].minWeight);
				from = dp[from][j].to;
			}
			k = k >> 1;
			j++;
		}
		cout << s << ' ' << m << endl;
	}

	return 0;
}

File: codeforces/702E/main.hs
-- ID:	702E (Analysis of Pathes in Functional Graph)
-- URL:	http://codeforces.com/contest/702/problem/E

-- This is not an accepted code!

import Control.Monad
import Data.Maybe
import Data.Array
import Data.Int
import Data.Char
import Numeric
import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BS8

getInt = read `fmap` getLine :: IO Int
getInts = (map (fst . fromJust . BS8.readInt) . BS8.words) `fmap` BS.getLine :: IO [Int]
getInt64s = (map (fst . fromJust . BS8.readInteger) . BS8.words) `fmap` BS.getLine :: IO [Int64]

-- (to, cost, minWeight)
type State = Array Int (Int, Int64, Int64)

toBinary :: Int64 -> String
toBinary x = showIntAtBase 2 intToDigit x ""

main = do
	[n,k] <- getInt64s :: IO [Int64]
	let pos = toBinary k
	nexts <- listArray (0,n-1) `fmap` getInt64s :: IO (Array Int64 Int64)
	costs <- listArray (0,n-1) `fmap` getInt64s :: IO (Array Int64 Int64)
	let state0 = initState n nexts costs
	let states = scanl (step n) state0 [1..length pos-1]
	let valids = pos `seq` fst $ unzip $ filter valid (zip states (reverse pos))
	let soln = findSoln valids n
	-- printing solutions seems not the problem
	forM_ soln $ \(_, si, mi) -> do
		putStr (show si)
		putChar ' '
		putStrLn (show mi)

--findSoln :: [State] -> Int64 -> [(Int64,Int64,Int64)]
findSoln :: [State] -> Int64 -> Array Int64 (Int64,Int64,Int64)
findSoln states n = f states $ listArray (0,n-1) [(i,0,10^8+1) | i <- [0..n-1]] where
	f [] soln = soln
	f (st:states) curr = f states $
		listArray (0,n-1) [(to1,s+len1, min m min1)
							| i <- [0..n-1], let (to,s,m) = curr ! i,
							let (to1,len1,min1) = st ! to]
	--f (st:states) curr = f states [(to1,s+len1, min m min1) | (i,s,m) <- curr, let (to1,len1,min1)=st!i]

valid (state, digit) = digit == '1'

initState :: Int64 -> Array Int64 Int64 -> Array Int64 Int64 -> State
initState n nexts costs = listArray (0,n-1) [(next i, cost i, cost i) | i <- [0..n-1]] where
	cost i = costs ! i
	next i = nexts ! i

step :: Int64 -> State -> Int -> State
step n state _ = listArray (0,n-1) [(to2, len1+len2, min min1 min2) | i <- [0..n-1]] where
	(to1, len1, min1) = state ! i
	(to2, len2, min2) = state ! to1
